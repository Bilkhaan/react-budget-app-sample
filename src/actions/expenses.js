import { v4 as uuidv4 } from 'uuid';
import database from '../firebase/firebase';

export const addExpense = (expense = {}) => (
  {
    type: 'ADD_EXPENSE',
    expense
  }
);

export const startAddExpense = (expense = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = "",
      amount = 0,
      createdAt = 0  
    } = expense;
    const expenseData = { description, note, amount, createdAt };

    return database.ref('expenses').push(expenseData).then(
      (ref) => {
        dispatch(addExpense(
          {
            id: ref.key,
            ...expenseData
          }
        ));
      }
    );
  }
}

export const removeExpense = ({ id } = {}) => (
  {
    type: 'REMOVE_EXPENSE',
    id
  }
);

export const startRemoveExpense = ({ id }) => {
  console.log(id)
  return (dispatch) => {
    return database.ref(`expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({ id }));
    });
  }
}

export const editExpense = (id, updates) => (
  {
    type: 'EDIT_EXPENSE',
    id,
    updates
  }
);

export const startEditExpense = (id, updates) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id, updates));
    });
  }
}

export const setExpenses = (expenses) => (
  {
    type: 'SET_EXPENSES',
    expenses
  }
);

export const startSetExpenses = () => {
  return (dispatch) => {
    return database.ref('expenses').once('value').then((snapshot) => {
      const expenses = [];
      snapshot.forEach((snapshotElem) => {
        expenses.push({
          id: snapshotElem.key,
          ...snapshotElem.val()
        });
      })
      dispatch(setExpenses(expenses));
    });
  }
}