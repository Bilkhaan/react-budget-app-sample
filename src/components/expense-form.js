import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
// import 'react-dates/initialize';

export class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense && props.expense.description || '',
      amount: props.expense && props.expense.amount/100 || '',
      note: props.expense && props.expense.note || '',
      createdAt: props.expense && moment(props.expense.createdAt) || moment(),
      calendarFocused: false,
      error: ''
    }

  }

  setDescription = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  }

  setAmount = (e) => {
    const amount = e.target.value;

    if (amount.match(/^\d*(\.\d{0,2})?$/))
      this.setState(() => ({ amount }));
  }

  onDateChange = (createdAt) => {
    this.setState(() => ({ createdAt }));
  }

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  }

  setNote = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  }

  onSubmit = (e) => {
    e.preventDefault();
  
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please fill in descripton and amount' }));
    } else {
      this.setState(() => ({ error: '' }));

      this.props.onsubmitExpense({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  }

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            id="description"
            placeholder="Description"
            value={this.state.description}
            onChange={this.setDescription}
          />
          <input
            type="text"
            placeholder="Amount"
            id="amount"
            value={this.state.amount}
            onChange={this.setAmount}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add note (optional)"
            id="note"
            value={this.state.note}
            onChange={this.setNote}
          >
          </textarea>

          <button >Submit Expense</button>
        </form>
      </div>
    );
  }
}