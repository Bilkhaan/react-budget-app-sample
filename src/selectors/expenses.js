import moment from 'moment';

export const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses.filter((expense) => {
    const momentCreatedAt = moment(expense.createdAt);
    const startDateMatched = startDate ?  startDate.isSameOrBefore(momentCreatedAt) : true;
    const endDateMatched = endDate ? endDate.isSameOrAfter(momentCreatedAt) : true;
    const textMatched = expense.description.toLowerCase().includes(text.toLowerCase())

    console.log(startDateMatched, endDateMatched, textMatched)
    return startDateMatched && endDateMatched && textMatched;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'amount') {
      return a.amount < b.amount ? -1 : 1;
    }
  })
}