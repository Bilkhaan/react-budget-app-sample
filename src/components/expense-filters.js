import React from 'react';
import { connect } from 'react-redux';
import { updateTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';
// import 'react-dates/initialize';
// import 'react-dates/lib/css/_datepicker.css';

export class ExpenseFilters extends React.Component {
  state = {
    calendarFocused: null
  }

  onDatesChange = ({startDate, endDate}) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  }

  onFocusChange = (calendarFocused) => {
    this.setState(() => ({calendarFocused}));
  }

  onTextChange = (e) => {
    this.props.updateTextFilter(e.target.value);
  }

  onSortByChange = (e) => {
    const sortBy = e.target.value;
    if (sortBy == 'amount')
      this.props.sortByAmount();
    else if (sortBy == 'date')
    this.props.sortByDate();
  }

  render() {
    return (
      <div>
        <input id="text-filter" type="text" value={this.props.filters.text} onChange={this.onTextChange}/>
        <select id="sort-filter" value={this.props.filters.sortBy} onChange={this.onSortByChange}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
          startDateId="your_unique_start_date_id"
          endDateId="enddate-id"
        />
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => ({
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  updateTextFilter: (text) => dispatch(updateTextFilter(text)),
  sortByAmount: () => dispatch(sortByAmount()),
  sortByDate: () => dispatch(sortByDate())
})

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  }
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseFilters);