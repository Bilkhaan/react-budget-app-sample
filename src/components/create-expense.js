import React from 'react';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';
import { ExpenseForm } from './expense-form'

export class CreateExpense extends React.Component {
  onSubmit = (expense) => {
    this.props.onSubmit(expense);
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
      this is Create view
      <ExpenseForm 
        onsubmitExpense={this.onSubmit}
      />
    </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    onSubmit: (expense) => dispatch(addExpense(expense))
  };
};

export default connect(undefined, mapDispatchToProps)(CreateExpense);