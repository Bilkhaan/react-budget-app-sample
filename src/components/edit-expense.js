import React from 'react';
import { connect } from 'react-redux';
import { ExpenseForm } from './expense-form';
import { editExpense, removeExpense } from '../actions/expenses';

export class EditExpense extends React.Component {
  editExpense = (expense) => {
    this.props.editExpense(this.props.match.params.id, expense);
    this.props.history.push('/');
  }

  onRemoveClicked = () => {
    this.props.removeExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  }
  
  render() {
    return (
      <div>
      this is Edit view
      <ExpenseForm
        expense={this.props.expense}
        onsubmitExpense={this.editExpense}
      />
      <button
        id="remove-expense"
        onClick={this.onRemoveClicked}
      >
        Remove
      </button>
    </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  const id = props.match.params.id;
  return {
    expense: state.expenses.find((expense) => {
      return id == expense.id;
    })
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: (id) => dispatch(removeExpense({ id }))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpense);