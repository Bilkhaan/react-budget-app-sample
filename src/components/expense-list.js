import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './expense-list-item'
import { getVisibleExpenses } from '../selectors/expenses';

export const ExpenseList = (props) => (
  <div>
    <h1>Expense List</h1>
    {
      !props.expenses || props.expenses.length == 0 ? 
        (<p>No expenses available</p>) : 
        (props.expenses.map((expense) => (<ExpenseListItem key={expense.id} {...expense} />)))
    }
  </div>
);

const mappedStateToProps = (state) => {
  return {
    expenses: getVisibleExpenses(state.expenses, state.filters)
  }
};

export default connect(mappedStateToProps)(ExpenseList);
