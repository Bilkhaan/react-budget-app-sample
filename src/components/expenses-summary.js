import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import expenseTotal from '../selectors/expense-total';
import {getVisibleExpenses} from '../selectors/expenses';

export const ExpenseSummary = (props) => {
  const expenseWord = props.expenseCount == 1 ? 'expense' : 'expenses';
  const expenseTotalFormated = numeral(props.expenseTotal).format('$0,0.00');

  return (
    <div>
      <h3>
      {`Showing ${props.expenseCount} ${expenseWord} totalling to ${expenseTotalFormated}`}
      </h3>
    </div>
  );
}

const mapStateToProp = (state) => {
  const visibleExpenses = getVisibleExpenses(state.expenses, state.filters);

  return {
    expenseCount: visibleExpenses.length,
    expenseTotal: expenseTotal(visibleExpenses)
  }
}

export default connect(mapStateToProp)(ExpenseSummary);
