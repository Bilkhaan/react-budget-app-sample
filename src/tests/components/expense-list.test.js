import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/expense-list';
import expenses from '../fixtures/expenses';

test('Should test expense list with data', () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />);
  expect(wrapper).toMatchSnapshot();
});

test('Should test expense list with no data', () => {
  const wrapper = shallow(<ExpenseList />);
  expect(wrapper).toMatchSnapshot();
});