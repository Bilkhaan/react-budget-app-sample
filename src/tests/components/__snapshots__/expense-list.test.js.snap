// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should test expense list with data 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <ExpenseListItem
    amount={2300}
    createdAt={0}
    description="Rent bill"
    id="1"
    key="1"
    note="Last month rent"
  />
  <ExpenseListItem
    amount={220}
    createdAt={10333234}
    description="Coffee bill"
    id="2"
    key="2"
    note="Last month rent"
  />
  <ExpenseListItem
    amount={1200}
    createdAt={-345600000}
    description="Cable bill"
    id="3"
    key="3"
    note="Last month rent"
  />
  <ExpenseListItem
    amount={3200}
    createdAt={345600000}
    description=" Car bill"
    id="4"
    key="4"
    note="Last month rent"
  />
</div>
`;

exports[`Should test expense list with no data 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <p>
    No expenses available
  </p>
</div>
`;
